lib/ssl/src/tls_handshake_1_3.erl

```diff
@@ -804,42 +805,36 @@ update_encryption_state(client, State) ->
-                           SslOptions, CRLDbHandle, Role, Host, OcspState0) ->
-    {Certs, CertExt, OcspState} = split_cert_entries(CertEntries, OcspState0),
-    OcspResponderCerts =
-        case maps:get(ocsp_stapling, SslOptions, disabled) of
-            #{ocsp_responder_certs := V} ->
-                V;
-            disabled ->
-                ?DEFAULT_OCSP_RESPONDER_CERTS
-        end,
-    ssl_handshake:certify(#certificate{asn1_certificates = Certs}, CertDbHandle, CertDbRef,
-                          SslOptions, CRLDbHandle, Role, Host, ?TLS_1_3,
-                          #{cert_ext => CertExt,
-                            ocsp_state => OcspState,
-                            ocsp_responder_certs => OcspResponderCerts}).
+                           SslOptions, CRLDbHandle, Role, Host, StaplingState0) ->
+    {Certs, StaplingInfo} = stapling_info(CertEntries, StaplingState0),
+    ssl_handshake:certify(#certificate{asn1_certificates = Certs}, CertDbHandle,
+                          CertDbRef, SslOptions, CRLDbHandle, Role, Host, ?TLS_1_3,
+                          StaplingInfo).
 
 store_peer_cert(#state{session = Session,
                        handshake_env = HsEnv} = State, PeerCert, PublicKeyInfo) ->
     State#state{session = Session#session{peer_certificate = PeerCert},
                 handshake_env = HsEnv#handshake_env{public_key_info = PublicKeyInfo}}.
 
-split_cert_entries(CertEntries, OcspState) ->
-    split_cert_entries(CertEntries, OcspState, [], #{}).
+stapling_info(CertEntries, StaplingState) ->
+    split_cert_entries(CertEntries, StaplingState, [], #{}).
 
-split_cert_entries([], OcspState, Chain, Ext) ->
-    {lists:reverse(Chain), Ext, OcspState};
-split_cert_entries([#certificate_entry{data = DerCert, extensions = Extensions0} | CertEntries],
-                   OcspState0, Chain, Ext) ->
+split_cert_entries([], StaplingState, Chain, CertExt) ->
+    {lists:reverse(Chain), #{cert_ext => CertExt,
+                             stapling_state => StaplingState}};
+split_cert_entries([#certificate_entry{data = DerCert,
+                                       extensions = Extensions0} | CertEntries],
+                   StaplingState0, Chain, CertExt) ->
     Id = public_key:pkix_subject_id(DerCert),
     Extensions = [ExtValue || {_, ExtValue} <- maps:to_list(Extensions0)],
-    OcspState = case maps:get(status_request, Extensions0, undefined) of
-                    undefined ->
-                        OcspState0;
-                    _ ->
-                        OcspState0#{ocsp_expect => stapled}
-                end,
-    split_cert_entries(CertEntries, OcspState, [DerCert | Chain], Ext#{Id => Extensions}).
+    StaplingState = case maps:get(status_request, Extensions0, undefined) of
+                        undefined ->
+                            StaplingState0;
+                        _ ->
+                            StaplingState0#{status => stapled}
```

Comment:
code here seems to not care much about which certificate had a staple nor about if all certificates got a staple or only one.
this might be problematic, if for example we receive a staple for non peer certificate and treat the whole path as not revoked and OK.

Reply:
I think this code determines that there is at least on staple  and all present staples will be verified in the path validation. They are saved in the state for this purpose.

Reply:
Do we need to be able to verify that the whole chain has been stapled? If so should be done during path validation 
in verify callback code in ssl_certificate.

Reply:
f983d0b909 adjusts that. handshake will fail if stapling is configured but peer(server) cert does not arrive with a valid status extension.

Modifying it to verifying whole chain should be easy to achieve, but currently we are strict only about peer cert.